src/FffGcodeWriter.cpp

```diff
@@ -1613,58 +1617,76 @@ bool FffGcodeWriter::processSingleLayerInfill(const SliceDataStorage& storage, L
+        infill_comp.generate(wall_tool_paths.back(), infill_lines, mesh.cross_fill_provider, &mesh);
+    }
+
+    const bool walls_generated = std::any_of(wall_tool_paths.cbegin(), wall_tool_paths.cend(), [](ToolPaths tp){ return !tp.empty(); });
+    if (!infill_lines.empty() || walls_generated)
     {
         added_something = true;
         setExtruder_addPrime(storage, gcode_layer, extruder_nr);
         gcode_layer.setIsInside(true); // going to print stuff inside print object
         std::optional<Point> near_start_location;
         if (mesh.settings.get<bool>("infill_randomize_start_location"))
         {
-            srand(gcode_layer.getLayerNr());
+            srand(static_cast<unsigned int>(gcode_layer.getLayerNr()));
             if(!infill_lines.empty())
             {
-                near_start_location = infill_lines[rand() % infill_lines.size()][0];
+                near_start_location = infill_lines[static_cast<unsigned int>(rand()) % infill_lines.size()][0];
             }
             else
             {
-                PolygonRef start_poly = infill_polygons[rand() % infill_polygons.size()];
-                near_start_location = start_poly[rand() % start_poly.size()];
+                // TODO: libArachne randomize infill walls
+                PolygonRef start_poly = infill_polygons[static_cast<unsigned int>(rand()) % infill_polygons.size()];
+                near_start_location = start_poly[static_cast<unsigned long>(rand()) % start_poly.size()];
             }
         }
-        if (!infill_polygons.empty())
+        if (walls_generated)
         {
-            constexpr bool force_comb_retract = false;
-            // start the infill polygons at the nearest vertex to the current location
-            gcode_layer.addTravel(PolygonUtils::findNearestVert(gcode_layer.getLastPlannedPositionOrStartingPosition(), infill_polygons).p(), force_comb_retract);
-            gcode_layer.addPolygonsByOptimizer(infill_polygons, mesh_config.infill_config[0], ZSeamConfig(), 0, false, 1.0_r, false, false, near_start_location);
+            for(const ToolPaths& tool_paths: wall_tool_paths)
+            {
+                BinJunctions bins = WallToolPaths::toolPathsToBinJunctions(tool_paths, wall_line_count);
+                for (const PathJunctions& paths : bins)
+                {
+                    for (const LineJunctions& line : paths)
+                    {
+//                      Todo: do a proper translation of these commands to the addWall
+//                      gcode_layer.addTravel(PolygonUtils::findNearestVert(gcode_layer.getLastPlannedPositionOrStartingPosition(), infill_polygons).p(), force_comb_retract);
+//                      gcode_layer.addPolygonsByOptimizer(infill_polygons, mesh_config.infill_config[0], ZSeamConfig(), 0, false, 1.0_r, false, false, near_start_location);
+                        gcode_layer.addWall(line, mesh_config.infill_config[0], false);
+                    }
+                }
```

Comment:
It would make more sense to let it add the polygons by optimizer again (but then in a way that's compatible with these variable width paths).
The logic that dictates that lines should be printed in indexed order should be contained in the `PathOrderOptimizer` which chooses between indexed ordering and the to-be-implemented optimized ordering by itself. Having logic regarding these "bins" outside of that class would be bad design. Let that class convert the toolpaths to these bins if it needs to, and otherwise not.